src/bm_sim/calculations.cpp

```diff
@@ -360,36 +360,110 @@ struct csum16 {
   }
 };
 
+// Concatenate the next n bytes into the specified type.
+// Pad zeros at tail when given bytes are shorter than the type.
+// For example, concat<uint32_t>(&{0xab, 0xcd}, 2) == 0xabcd0000.
+template <typename T>
+T concat(const char *buf, size_t n) {
+  static constexpr size_t type_size = sizeof(T);
+  assert(n <= type_size);
+  T res = 0;
+  if (n > 0) {
+    for (size_t offset = 0; offset < n; offset++) {
+      res = (res << 8) | (static_cast<T>(buf[offset]) & 0xff);
+    }
+    res <<= ((type_size - n) * 8);
+  }
+  return res;
+}
+
 template <typename T>
 struct xor_hash {
   T operator()(const char *buf, size_t len) const {
     static constexpr size_t n = sizeof(T);
     T final_xor_value = 0;
-    T mask = 0xff;
     size_t byte = 0;
-    T t;
 
     /* Main loop - n bytes at a time */
     while (len >= n) {
-      t = 0;
-      for (size_t offset = 0; offset < n; offset++) {
-        t = (t << 8) + (static_cast<T>(buf[byte + offset]) & mask);
-      }
-      final_xor_value ^= t;
+      final_xor_value ^= concat<T>(buf + byte, n);
       byte += n;
       len -= n;
     }
 
     /* Handle tail less than n bytes long */
-    t = 0;
-    for (size_t offset = 0; offset < len; offset++) {
-      t = (t << 8) + (static_cast<T>(buf[byte + offset]) & mask);
+    final_xor_value ^= concat<T>(buf + byte, len);
+
+    return final_xor_value;
+  }
+};
+
+struct toeplitz {
+  using rss_key_t = RssMgr::rss_key_t;
+
+  toeplitz() {
+    update_key({0x00, 0x00, 0x00, 0x00});
+  }
+
+  uint32_t operator()(const char *buf, size_t len) const {
+    std::unique_lock<std::mutex> lock(m);
+
+    // Convert input bytes (in network byte order) to host words.
+    // Pad the last one with zeros if it's shorter than a word.
+    std::vector<uint32_t> input_words;
+    auto raw_input_words = reinterpret_cast<const uint32_t *>(buf);
+    const size_t n_words = len / 4;
+    const size_t n_bytes_remain = len % 4;
+    for (size_t i_word = 0; i_word < n_words; i_word++) {
+      input_words.push_back(ntohl(raw_input_words[i_word]));
+    }
+    if (n_bytes_remain) {
+      input_words.push_back(
+        concat<uint32_t>(buf + (len - n_bytes_remain), n_bytes_remain));
+    }
```

Comment:
it doesn't seem that this needs to be protected by the mutex

Reply:
I see. I'll remove it then. I didn't truly understand why the mutex was used in `custom_crc` and simply followed their examples here.

Reply:
to be clear, you need a mutex since another thread can come it and mutate the rss_key
but this specific section of the code (building the `input_words`) is expensive (vector memory allocation) and doesn't need to be protected by the mutex
the mutex is only needed to access rss_key

Reply:
I see. Thanks for your clarification!